athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/StorageMetadata.java

```diff
@@ -0,0 +1,315 @@
+    {
+        requireNonNull(bucketName, "bucketName was null");
+        requireNonNull(path, "fileName was null");
+        LOGGER.info("Retrieving field schema from file {}, under the bucket {}", path, bucketName);
+        String uri = createUri(bucketName, path);
+        DatasetFactory factory = new FileSystemDatasetFactory(allocator,
+                NativeMemoryPool.getDefault(), format, uri);
+        // inspect schema
+        return factory.inspect();
+    }
+
+    private boolean canIncludePath(String folderPath, List<EqualsExpression> expressions)
+    {
+        if (expressions.isEmpty()) {
+            return true;
+        }
+        String[] folderPaths = folderPath.split("/");
+        for (String path : folderPaths) {
+            Optional<FieldValue> optionalFieldValue = FieldValue.from(path);
+            if (optionalFieldValue.isEmpty()) {
+                continue;
+            }
+            FieldValue fieldValue = optionalFieldValue.get();
+            Optional<EqualsExpression> optionalExpression = expressions.stream()
+                    .filter(expr -> expr.columnName().equalsIgnoreCase(fieldValue.getField()))
+                    .findFirst();
+            if (optionalExpression.isPresent()) {
+                LOGGER.info("Evaluating field value {} against the expression {}", fieldValue, expressions);
+                EqualsExpression expression = optionalExpression.get();
+                if (!expression.apply(fieldValue.getValue())) {
+                    return false;
+                }
+            }
+            else {
+                LOGGER.info("No expression found for field {}", fieldValue.getField());
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Builds the table schema based on the provided field
+     *
+     * @param table      Glue table object
+     * @return An instance of {@link Schema}
+     */
+    public Schema buildTableSchema(Table table, BufferAllocator allocator) throws URISyntaxException
+    {
+        SchemaBuilder schemaBuilder = SchemaBuilder.newBuilder();
+        String locationUri = table.getStorageDescriptor().getLocation();
+        URI storageLocation = new URI(locationUri);
+        List<Field> fieldList = getFields(storageLocation.getAuthority(), storageLocation.getPath(), table.getParameters().get(CLASSIFICATION_GLUE_TABLE_PARAM), allocator);
+
+        LOGGER.debug("Schema Fields\n{}", fieldList);
+        for (Field field : fieldList) {
+            if (isFieldTypeNull(field)) {
+                field = Field.nullable(field.getName().toLowerCase(), Types.MinorType.VARCHAR.getType());
+            }
+            else {
+                field = new Field(field.getName().toLowerCase(), new FieldType(field.isNullable(), field.getType(), field.getDictionary(), field.getMetadata()), field.getChildren());
+            }
+            schemaBuilder.addField(getCompatibleField(field));
```

Comment:
Why are we changing the user's provided Glue Table schema here?
If they want nanos and its not supported, we shouldn't just convert it to millis for them. They specified nanos as their schema, you can't just change their schema for them.

In this situation, you just have to throw an exception saying that we don't support nanos (or whatever other types we don't support).

The only time you can perform coercion is if the type isn't being explicitly provided by the customer.

Reply:
we are not changing user's provided Glue Table schema. 

we are changing/converting to Athena-supported datatypes, only when the user has not provided schema in GlueTable and we are detecting it from one of the files with the help of Dataset API.

Reply:
Sorry, I see now, the code is only using the Glue `Table` for the location URI and classification parameter.